release-notes/3.0/3.0.0-preview3.md

```diff
@@ -0,0 +1,176 @@
+  - Attempts authentication to Azure Active Directory using a managed identity that has been assigned to the deployment environment. **"Client Id" of "User Assigned Managed Identity"** shall be read from **"User Id" connection property**.
+- **SharedTokenCacheCredential**
+  - Authenticates using tokens in the local cache shared between Microsoft applications.
+- **VisualStudioCredential**
+  - Enables authentication to Azure Active Directory using data from Visual Studio
+- **VisualStudioCodeCredential**
+  - Enables authentication to Azure Active Directory using data from Visual Studio Code.
+- **AzureCliCredential**
+  - Enables authentication to Azure Active Directory using Azure CLI to obtain an access token.
+
+> InteractiveBrowserCredential is disabled by the driver implementation of "Active Directory Default", and "Active Directory Interactive" continues to be the only method to acquire token using MFA/Interactive authentication.*
+
+> Further customization options are not available at the moment.
+
+### Custom key store provider registration enhancements
+
+Microsoft.Data.SqlClient now enables support for multi-user custom key store provider registration for performing column encryption/decryption. Below APIs are introduced to allow registration of custom key store providers on an instance of `SqlConnection` and `SqlCommand`:
+
+```cs
+public class SqlConnection
+{
+        public void RegisterColumnEncryptionKeyStoreProvidersOnConnection(IDictionary<string, SqlColumnEncryptionKeyStoreProvider> customProviders)
+}
+public class SqlCommand 
+{
+        public void RegisterColumnEncryptionKeyStoreProvidersOnCommand(IDictionary<string, SqlColumnEncryptionKeyStoreProvider> customProviders)
+}
+```
+
+The static API on `SqlConnection`, i.e. `SqlConnection.RegisterColumnEncryptionKeyStoreProviders` to register key store providers globally continues to be supported. The column encryption key cache maintained globally only applies to globally registered providers.
+
+#### Key Store Provider cache precedence
+
+System key store providers will be used from the global system cache if the requested provider is one. No providers should be registered on the connection or command instances if a system provider is needed.
+Registered key store providers are cached by the driver in 3 different cache layers. The global layer is as-is. The new per-connection and per-command level caches will be empty initially and can be set more than once.
+
+The precedence of the three caches will be as follows:
+
+- The per-command cache will be checked if it is not empty.
+- If the per-command cache is empty, the per-connection cache will be checked if it is not empty.
+- If the per-connection cache is empty, the global cache will be checked.
+
+Once a cache is found, the driver will **NOT** fall back to the other caches to search for requested provider.
+An exception will be thrown containing only the registered providers in the cache that was checked.
```

Comment:
What exactly is a "system" provider? I find the terms global and system confusing without some clarification. I assume "system" refers to providers for which there is only one possible store on the computer running the program.

Here's my suggestion, but I may have some misunderstandings:
```suggestion
System key store providers are available for the Windows Certificate Store and ???. They will be used from the global registration if the requested provider is a system provider. No providers should be registered on the connection or command instances if a system provider is needed.
Key store providers can be registered with the driver at three different layers. The global level is as it currently is. The new per-connection and per-command level registrations will be empty initially and can be set more than once.

The precedence of the three registrations are as follows:

- The per-command registration will be checked if it is not empty.
- If the per-command registration is empty, the per-connection registration will be checked if it is not empty.
- If the per-connection registration is empty, the global registration will be checked.

Once any key store provider is found at a registration level, the driver will **NOT** fall back to the other registrations to search for a provider.
If providers are registered but the proper provider is not found at a level, an exception will be thrown containing only the registered providers in the registration that was checked.
```

Reply:
Yes, System Providers are Windows-only, available via application host OS only. I'll change the term to "built-in column master key store providers" (as per MS docs)